Rpc/Speaker.cs

```diff
@@ -0,0 +1,246 @@
+    public async Task StartAsync(CancellationToken ct = default)
+    {
+        // Handshakes should always finish quickly, so enforce a 5s timeout.
+        using var cts = CancellationTokenSource.CreateLinkedTokenSource(ct, _cts.Token);
+        cts.CancelAfter(TimeSpan.FromSeconds(5));
+        await PerformHandshake(ct);
+
+        // Start ReceiveLoop in the background.
+        _receiveTask = ReceiveLoop(_cts.Token);
+        _ = _receiveTask.ContinueWith(t =>
+        {
+            if (t.IsFaulted) Error?.Invoke(t.Exception!);
+        }, CancellationToken.None);
+    }
+
+    private async Task PerformHandshake(CancellationToken ct = default)
+    {
+        // Simultaneously write the header string and read the header string in
+        // case the conn is not buffered.
+        var headerCts = CancellationTokenSource.CreateLinkedTokenSource(ct, _cts.Token);
+        var writeTask = WriteHeader(headerCts.Token);
+        var readTask = ReadHeader(headerCts.Token);
+        await TaskUtilities.CancellableWhenAll(headerCts, writeTask, readTask);
+
+        var header = RpcHeader.Parse(await readTask);
+        var expectedRole = RpcMessage<TR>.GetRole();
+        if (header.Role != expectedRole)
+            throw new ArgumentException($"Expected peer role '{expectedRole}' but got '{header.Role}'");
+
+        header.Version.Validate(ApiVersion.Current);
+    }
+
+    private async Task WriteHeader(CancellationToken ct = default)
+    {
+        var header = new RpcHeader(RpcMessage<TS>.GetRole(), ApiVersion.Current);
+        await _conn.WriteAsync(header.ToBytes(), ct);
+    }
+
+    private async Task<string> ReadHeader(CancellationToken ct = default)
+    {
+        var buf = new byte[256];
+        var have = 0;
+
+        while (true)
+        {
+            // Read into buf[have:have+1]
+            await _conn.ReadExactlyAsync(buf, have, 1, ct);
+            if (buf[have] == '\n') break;
+            have++;
+            if (have >= buf.Length)
+                throw new IOException($"Header malformed or too large: '{Encoding.UTF8.GetString(buf)}'");
+        }
+
+        return Encoding.UTF8.GetString(buf, 0, have);
+    }
+
+    private async Task ReceiveLoop(CancellationToken ct = default)
+    {
+        try
+        {
+            while (!ct.IsCancellationRequested)
+            {
+                var message = await _serdes.ReadMessage(_conn, ct);
+                if (message.RpcField.ResponseTo != 0)
+                    // Look up the TaskCompletionSource for the message ID and
+                    // complete it with the message.
+                    if (_pendingReplies.TryRemove(message.RpcField.ResponseTo, out var tcs))
+                        tcs.SetResult(message);
+
+                // TODO: we should log unknown replies
+                // Start a new task in the background to handle the message.
+                _ = Task.Run(() => Receive?.Invoke(new ReplyableRpcMessage<TS, TR>(this, message)), ct);
```

Comment:
If I'm reading this correctly, if we get a reply message we _both_ return it to the requestor and trigger the receive event. Is that really what we want? The event handler would have to check again to make sure we're not also treating it as an unsolicited message.

Reply:
Yeah this was a mistake, I've added a `continue` to avoid writing replies to the event handler